Core/include/Acts/AmbiguityResolution/AthenaAmbiguityResolution.ipp

```diff
@@ -0,0 +1,406 @@
+    auto srcProxy = tracks.getTrack(iTrack);
+    destProxy.copyFrom(srcProxy, false);
+    destProxy.tipIndex() = srcProxy.tipIndex();
+  }
+
+  const TrackContainer<track_container_t, traj_t, holder_t> outputTracks{
+      std::make_shared<VectorTrackContainer>(std::move(*trackContainer)),
+      trackStateContainer};
+  return outputTracks;
+}
+
+template <typename track_container_t, typename traj_t,
+          template <typename> class holder_t, typename source_link_hash_t,
+          typename source_link_equality_t>
+std::vector<std::vector<std::tuple<std::size_t, std::size_t, bool>>>
+AthenaAmbiguityResolution::computeInitialState(
+    const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
+    source_link_hash_t&& sourceLinkHash,
+    source_link_equality_t&& sourceLinkEquality) const {
+  auto measurementIndexMap =
+      std::unordered_map<SourceLink, std::size_t, source_link_hash_t,
+                         source_link_equality_t>(0, sourceLinkHash,
+                                                 sourceLinkEquality);
+
+  std::vector<std::vector<std::tuple<std::size_t, std::size_t, bool>>>
+      measurementsPerTrack;
+
+  ACTS_VERBOSE("Starting to compute initial state");
+
+  for (const auto& track : tracks) {
+    std::vector<std::tuple<std::size_t, std::size_t, bool>> measurements_tuples;
+
+    for (auto ts : track.trackStatesReversed()) {
+      ACTS_DEBUG("Track state type check: ");
+      if (ts.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag)) {
+        Acts::SourceLink sourceLink = ts.getUncalibratedSourceLink();
+        ACTS_DEBUG("Track state type is MeasurementFlag");
+        const auto& geoID = ts.referenceSurface().geometryId();
+
+        // assign a new measurement index if the source link was not seen yet
+        auto emplace = measurementIndexMap.try_emplace(
+            sourceLink, measurementIndexMap.size());
+
+        bool isoutliner =
+            ts.typeFlags().test(Acts::TrackStateFlag::OutlierFlag);
+
+        measurements_tuples.push_back(
+            std::make_tuple(emplace.first->second, geoID.volume(), isoutliner));
+      }
+    }
+
+    measurementsPerTrack.push_back(std::move(measurements_tuples));
+  }
+
+  return measurementsPerTrack;
+}
+
+template <typename track_container_t, typename traj_t,
+          template <typename> class holder_t>
+std::vector<double> Acts::AthenaAmbiguityResolution::simpleScore(
+    const TrackContainer<track_container_t, traj_t, holder_t>& tracks,
+    std::vector<std::map<std::size_t, TrackFeatures>>& trackFeaturesMaps,
+    Optional_cuts<track_container_t, traj_t, holder_t> optionalCuts) const {
+  std::vector<double> trackScore;
+
+  int iTrack = 0;
+
+  ACTS_VERBOSE("Number of detectors: " << m_cfg.detectorMap.size());
+
+  ACTS_INFO("Starting to score tracks");
+
+  // Loop over all the tracks in the container
+  for (const auto& track : tracks) {
+    auto trackFeaturesMap = std::map<std::size_t, TrackFeatures>();
```

Comment:
can you remind me why the wave multiple (and independent) maps, one for each track? If these are totally independent, then why can't we process a track at a time and then use the same objects/memory for the other tracks?

Reply:
The TrackFeatureMap has been changed to TrackFeatureVector